docs/architecture/adr-006-light-client-concurrency.md

```diff
@@ -0,0 +1,106 @@
+# ADR 006: Light Client Concurrency
+
+## Changelog
+- 2020-03-15: Initial draft.
+
+## Context
+
+The light client protocol provides a method for verifying application
+state without the execution of all preceding transactions. The
+protocol for the light client is described in
+[English](https://github.com/tendermint/spec/tree/bucky/light-reorg/spec/consensus/light)
+while the rust implementation is described in
+[ADR-002](adr-002-light-client-adr-index.md). This ADR outlines the
+next iteration of the light client implementation in rust in which
+multiple modules (Peer Exchange, Requester, Verifier, etc) operate
+concurrently.
+
+The basis for this work comes from the learnings of the [Reactor
+Experiments](https://github.com/informalsystems/reactor-experiments) as
+well as the [Blockchain Reactor
+Refactor](https://github.com/tendermint/tendermint/blob/master/docs/architecture/adr-043-blockchain-riri-org.md). The key take aways from that work are:
+
+1. Communicate with events and not state
+2. Separate concerns into independently verifiable (by humans as well as
+   computers) components.
+3. Push IO to the edge of the system to allow simulation of components
+   as well as integration between components.
+
+## Decision
+
+Model the Peer Exchange, Requester, Detector and Verifier as separate
+independently verifiable components. Each component is a finite state
+machine with internal state. Communication between components is done by
+event dispatching to a routing layer. The routing layer handles input
+and output by routing to/from peers to components and between
+components. The routing layer will provide the necessary abstraction to
+allow simulation of individual components as well as integration between
+components and peers.
+
+![Concurrency Diagram](assets/light-client-concurrency.png)
+
+This architecture outlines components that have been 
+described in previous ADRs and will only be expanded upon here.
+
+### Peer Exchange
+The light clients requires an expanding set of peers to get headers
+from. The larger and more diverse the set of peers; the better the
+chances that the light client will detect a fork. The peer list can be
+seeded with peers from configuration and will can crawl the network.
+
+### Requester
+The light client requests validator sets and signed headers from peers.
+The order of data requested from peers is determined by the specific verification
+algorithm being followed. Specifically either sequential or bisection.
+Additionally the Requester is expected to request signed header for a
```

Comment:
_This may lead to mixed concerns between Requester and Detector modules. If you are trying to archive “better isolation of concerns”, it could be beneficial to leave the logic of picking what nodes to ask for headers to Detector module (also, when to ask for headers). This does not necessary mean Detector module would have to store a list of peers internally. It can use some kind of shortcuts (e.g. “all” - all backup nodes, “random” - random backup node, “randomN{N}” - random N backup nodes)_